### PR TITLE
7179006: [macosx] Print-to-file doesn't work: printing to the default printer instead.

### PR DESCRIPTION
**Pull request by prrace**
A long-standing bug on macOS: https://bugs.openjdk.java.net/browse/JDK-7179006

The fix is to propagate whatever is set as the Destination down to native and set it on the native printing object.

Also if using the native dialog, but with attributes, copy back the destination from native to the Java attribute set.

The new test can directly test printing to file in an automated way - so long as there is a printer.

The updated manual test can be used to verify the  cross-platform dialog case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-7179006](https://bugs.openjdk.java.net/browse/JDK-7179006): [macosx] Print-to-file doesn't work: printing to the default printer instead ⚠️ Title mismatch between PR and JBS.

### Reviewers
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/339/head:pull/339`
`$ git checkout pull/339`

### CODE REVIEW

**Comment by prrace** [comment]
label /remove awt

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m [hunk 1]
```diff
@@ -423,6 +429,19 @@ static void javaPrinterJobToNSPrintInfo(JNIEnv* env, jobject srcPrinterJob, jobj
...
+
+    jstring dest = JNFCallObjectMethod(env, srcPrinterJob, jm_getDestinationFile);
+    if (dest != NULL) {
+       NSString *nsDestStr = JNFJavaToNSString(env, dest);
+       [dst setJobDisposition:NSPrintSaveJob];
+      // [printingDictionary setObject:nsDestStr forKey:NSPrintSavePath];
+       //NSURL *nsURL = [NSURL URLWithString:nsDestStr];
+       NSURL *nsURL = [NSURL fileURLWithPath:nsDestStr isDirectory:NO];
+       [printingDictionary setObject:nsURL forKey:NSPrintJobSavingURL];
+//        JNFCallVoidMethod(env, dstPrinterJob, jm_setPrintToFile, true);
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m:441]: Does it look like the commented lines are not needed?

File: test/jdk/java/awt/print/PrinterJob/PrintToFileTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            return;
+        }
+        pj.setPrintable(new PrintToFileTest(), new PageFormat());
+        PrintRequestAttributeSet pSet = new HashPrintRequestAttributeSet();
+        File file = new File("./out.prn");
+        pSet.add(new Destination(file.toURI()));
+        pj.print(pSet);
+        if (!file.exists()) {
+             throw new RuntimeException("No file created");
+        }
```
Comment (prsadhuk) [inline test/jdk/java/awt/print/PrinterJob/PrintToFileTest.java:60]: I guess we need to delete the file in finally block, if created.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java [hunk 3]
```diff
@@ -251,6 +253,22 @@ private void setPrintToFile(boolean printToFile) {
         isPrintToFile = printToFile;
     }
 
+    private void setDestinationFile(String dest) {
+        System.out.println("dest="+dest+" attrs="+attributes);
```
Comment (prsadhuk) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:257]: Probably this println should be removed.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m (hunk 1 above)
Comment (prrace) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m:441]: Hmm. I could swear I removed those and re-ordered the first two lines as well.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java (hunk 3 above)
Comment (prrace) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:257]: oops

File: test/jdk/java/awt/print/PrinterJob/PrintToFileTest.java (hunk 2 above)
Comment (prrace) [reply test/jdk/java/awt/print/PrinterJob/PrintToFileTest.java:60]: I meant to leave it. So we can look at it if we need to.

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java [hunk 4]
```diff
@@ -251,6 +253,21 @@ private void setPrintToFile(boolean printToFile) {
         isPrintToFile = printToFile;
     }
 
+    private void setDestinationFile(String dest) {
+        if (attributes != null && dest != null) {
+            try {
+               URI destURI = new URI(dest);
+               attributes.add(new Destination(destURI));
+               destinationAttr = "" + destURI.getSchemeSpecificPart();
```
Comment (mrserb) [inline src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:261]: This destinationAttr in the RasterPrinterJob is usually assigned to:
                destinationAttr = "" + new File(destination.getURI().
                                                getSchemeSpecificPart());
Do we need to do the same here?

File: src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m [hunk 5]
```diff
@@ -316,21 +316,25 @@ static void nsPrintInfoToJavaPrinterJob(JNIEnv* env, NSPrintInfo* src, jobject d
...
         JNFCallVoidMethod(env, dstPrinterJob, jm_setPrintToFile, true);
+        NSURL *url = [printingDictionary objectForKey:NSPrintJobSavingURL];
+        NSString *nsStr = [url absoluteString];
+        jstring str = JNFNSToJavaString(env, nsStr);
+        JNFCallVoidMethod(env, dstPrinterJob, jm_setDestinationFile, str);
     } else {
         JNFCallVoidMethod(env, dstPrinterJob, jm_setPrintToFile, false);
     }
 
     // get the selected printer's name, and set the appropriate PrintService on the Java side
```
Comment (mrserb) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/CPrinterJob.m:337]: The comment need to be moved up as well?

File: src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java (hunk 4 above)
Comment (prrace) [reply src/java.desktop/macosx/classes/sun/lwawt/macosx/CPrinterJob.java:261]: I see that in WPrinterJob.java we do the same as I am doing here.
I think the difference in RPJ may bethat we want to get a full path name which is what the new File did.
Here macOS gives us a full path name already. So I don't think it is needed.
It gave the required value for sure.

**Comment by prrace** [comment]
Updated with the comment moved.

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**